### PR TITLE
Make CUDA device properties's `__repr__` output actually printable

### PR DESCRIPTION
**Pull request by int3**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132863

Previously we would write the UUID bytes directly, leading to 'invalid
UTF-8 sequence' errors.

### CODE REVIEW

File: torch/csrc/cuda/Module.cpp [hunk 1]
```diff
@@ -914,6 +914,40 @@ PyObject* THCPModule_cudaGetSyncDebugMode(PyObject* self, PyObject* noargs) {
...
+      (uint8_t)uuid_bytes[8],
+      (uint8_t)uuid_bytes[9],
+      (uint8_t)uuid_bytes[10],
+      (uint8_t)uuid_bytes[11],
+      (uint8_t)uuid_bytes[12],
+      (uint8_t)uuid_bytes[13],
+      (uint8_t)uuid_bytes[14],
+      (uint8_t)uuid_bytes[15]);
+  return std::string(uuid_chars.data());
+}
```
Comment (Skylion007) [inline torch/csrc/cuda/Module.cpp:949]:
```suggestion
std::string uuid_to_string(const char* uuid_bytes) {
  // UUIDs are a 128-bit label. CUDA and HIP store this as char[16].
  // For string representation, the code here expands this to
  // 8-4-4-4-12 hex format, so each byte becomes 2 hex characters.
  // Size is 16x2 hex characters + 4 hyphens.
  constexpr size_t size = 16 * 2 + 4 + 1;
  std::array<char, size> uuid_chars {};
  fmt::format_to(
    uuid_chars.data(),
    "{:02x}{:02x}{:02x}{:02x}-"
    "{:02x}{:02x}-"
    "{:02x}{:02x}-"
    "{:02x}{:02x}-"
    "{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}",
    (uint8_t) uuid_bytes[0],  
    (uint8_t) uuid_bytes[1],  
    (uint8_t) uuid_bytes[2],  
    (uint8_t) uuid_bytes[3],
    (uint8_t) uuid_bytes[4],  
    (uint8_t) uuid_bytes[5],  
    (uint8_t) uuid_bytes[6],  
    (uint8_t) uuid_bytes[7],
    (uint8_t) uuid_bytes[8],  
    (uint8_t) uuid_bytes[9],  
    (uint8_t) uuid_bytes[10], 
    (uint8_t) uuid_bytes[11],
    (uint8_t) uuid_bytes[12], 
    (uint8_t) uuid_bytes[13], 
    (uint8_t) uuid_bytes[14], 
    (uint8_t) uuid_bytes[15]
  );
  return std::string(uuid_chars.data());
}
```
use fmtlib to make this a bit more efficient, add an `include <fmt/core.h` you cna also try something like this which is even more readable:
```suggestion
std::string uuid_to_string(const char* uuid_bytes) {
  // UUIDs are a 128-bit label. CUDA and HIP store this as char[16].
  // For string representation, the code here expands this to
  // 8-4-4-4-12 hex format, so each byte becomes 2 hex characters.
  std::array<uint8_t, 16> bytes;
  std::copy(uuid_bytes, uuid_bytes + 16, bytes.begin());

  return fmt::format(
    "{:02x}{:02x}{:02x}{:02x}-"
    "{:02x}{:02x}-"
    "{:02x}{:02x}-"
    "{:02x}{:02x}-"
    "{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}",
    bytes[0], 
    bytes[1], 
    bytes[2], 
    bytes[3],
    bytes[4], 
    bytes[5], 
    bytes[6], 
    bytes[7],
    bytes[8], 
    bytes[9], 
    bytes[10], 
    bytes[11],
    bytes[12], 
    bytes[13], 
    bytes[14], 
    bytes[15]
  );
}
```
if you add add`include <algorithm>` and `include <fmt/ranges.h>`.

**Review by Skylion007 — APPROVED** [review]
If you switch to fmtlib and fix the lint issues looks good.

**Review by Skylion007 — APPROVED** [review]

**Review by eqy — APPROVED** [review]

**Pull request closed without merging**